File: core/src/main/java/com/tuuzed/androidx/recyclerview/adapter/RecyclerViewAdapter.java
package com.tuuzed.androidx.recyclerview.adapter;

import android.view.ViewGroup;

import com.tuuzed.androidx.recyclerview.adapter.internal.ItemTypePool;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import androidx.annotation.IntRange;
import androidx.annotation.LayoutRes;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

public final class RecyclerViewAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {
    @NonNull
    private List<Object> mItems;
    @NonNull
    private final ItemTypePool mItemTypePool;

    public RecyclerViewAdapter() {
        this(new ArrayList<>(), 16);
    }

    public RecyclerViewAdapter(@NonNull List<Object> items, @IntRange(from = 1) int initTypeCapacity) {
        mItems = items;
        this.mItemTypePool = new ItemTypePool(initTypeCapacity);
    }

    @Override
    public final int getItemViewType(int position) {
        Object item = mItems.get(position);
        return mItemTypePool.indexOf(item.getClass());
    }

    @NonNull
    @Override
    public final RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int indexViewType) {
        ItemTypeViewBinder<?, ?> itemTypeViewBinder = mItemTypePool.findItemTypeViewBinder(indexViewType);
        return itemTypeViewBinder.onCreateViewHolder(parent, indexViewType);
    }

    @Override
    public final void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, int position) {
        int indexViewType = getItemViewType(position);
        Object item = mItems.get(position);
        ItemTypeViewBinder binder = mItemTypePool.findItemTypeViewBinder(indexViewType);
        //noinspection unchecked
        binder.onBindViewHolder(holder, item, position);
    }

    @Override
    public int getItemCount() {
        return mItems.size();
    }


    @NonNull
    public <T, VH extends RecyclerView.ViewHolder> RecyclerViewAdapter bind(
            @NonNull final Class<T> itemType,
            @NonNull final ItemTypeViewBinder<T, VH> itemTypeViewBinder
    ) {
        mItemTypePool.register(itemType, itemTypeViewBinder);
        return this;
    }

    @NonNull
    public <T, VH extends RecyclerView.ViewHolder> ItemTypeViewBinder<T, VH> bind(
            @NonNull final Class<T> itemType,
            @NonNull final OnCreateViewHolder<VH> onCreateViewHolder,
            @NonNull final OnBindViewHolder<T, VH> onBindViewHolder
    ) {
        ItemTypeViewBinder<T, VH> binder = new ItemTypeViewBinder<T, VH>() {
            @Override
            public void onBindViewHolder(@NonNull VH holder, @NonNull T item, int position) {
                onBindViewHolder.onBindViewHolder(holder, item, position);
            }

            @NonNull
            @Override
            public VH onCreateViewHolder(ViewGroup parent, int viewType) {
                return onCreateViewHolder.onCreateViewHolder(parent, viewType);
            }
        };
        bind(itemType, binder);
        return binder;
    }

    @NonNull
    public <T> ItemTypeViewBinder<T, CommonItemViewHolder> bind(
            @NonNull final Class<T> itemType,
            @LayoutRes final int layoutId,
            @NonNull final OnBindViewHolder<T, CommonItemViewHolder> onBindViewHolder
    ) {
        ItemTypeViewBinder<T, CommonItemViewHolder> binder = new AbstractItemViewBinder<T>() {
            @Override
            public int getLayoutId() {
                return layoutId;
            }

            @Override
            public void onBindViewHolder(@NonNull CommonItemViewHolder holder, @NonNull T item, int position) {
                onBindViewHolder.onBindViewHolder(holder, item, position);
            }
        };
        bind(itemType, binder);
        return binder;
    }

    @NonNull
    public RecyclerViewAdapter with(@NonNull RecyclerView recyclerView) {
        recyclerView.setAdapter(this);
        return this;
    }

    @NonNull
    public List<Object> items() {
        return getItems();
    }

    @NonNull
    public List<Object> getItems() {
        return mItems;
    }

    @NonNull
    public RecyclerViewAdapter setItems(@NonNull List<Object> items) {
        this.mItems = items;
        return this;
    }


    @NonNull
    public RecyclerViewAdapter appendItems(@NonNull List<Object> items) {
        this.mItems.addAll(items);
        return this;
    }

    @NonNull
    public RecyclerViewAdapter appendItems(@NonNull Object... items) {
        Collections.addAll(mItems, items);
        return this;
    }
}

File: core/src/main/java/com/tuuzed/androidx/recyclerview/adapter/RecyclerViewAdapter.kt
package com.tuuzed.androidx.recyclerview.adapter

import android.view.ViewGroup
import androidx.annotation.IntRange
import androidx.annotation.LayoutRes
import androidx.recyclerview.widget.RecyclerView
import java.util.*

class RecyclerViewAdapter constructor(
        var items: MutableList<Any>,
        @IntRange(from = 1) initTypeCapacity: Int
) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {

    companion object {

        @JvmOverloads
        fun with(recyclerView: RecyclerView,
                 items: MutableList<Any> = ArrayList(),
                 @IntRange(from = 1) initTypeCapacity: Int = 16,
                 block: (RecyclerViewAdapter.() -> Unit)? = null
        ): RecyclerViewAdapter {
            return RecyclerViewAdapter(items, initTypeCapacity).also {
                it.with(recyclerView)
                if (block != null) {
                    it.with(recyclerView, block)
                }
            }
        }

    }

    private val mItemTypePool: ItemTypePool = ItemTypePool(initTypeCapacity)

    override fun getItemViewType(position: Int): Int {
        val item = items[position]
        return mItemTypePool.indexOf(item.javaClass)
    }

    override fun onCreateViewHolder(parent: ViewGroup, indexViewType: Int): RecyclerView.ViewHolder {
        val itemTypeViewBinder = mItemTypePool.findItemTypeViewBinder(indexViewType)
        return itemTypeViewBinder.onCreateViewHolder(parent, indexViewType)
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        val indexViewType = getItemViewType(position)
        val item = items[position]
        mItemTypePool.findItemTypeViewBinder(indexViewType).onBindViewHolder(holder, item, position)
    }

    override fun getItemCount(): Int = items.size

    fun <T, VH : RecyclerView.ViewHolder> bind(itemType: Class<in T>, itemTypeViewBinder: ItemTypeViewBinder<T, VH>): RecyclerViewAdapter {
        mItemTypePool.register(itemType, itemTypeViewBinder)
        return this
    }

    fun <T, VH : RecyclerView.ViewHolder> bind(itemType: Class<T>, onCreateViewHolder: OnCreateViewHolder<VH>, onBindViewHolder: OnBindViewHolder<T, VH>): ItemTypeViewBinder<T, VH> {
        val binder = object : ItemTypeViewBinder<T, VH> {
            override fun onBindViewHolder(holder: VH, item: T, position: Int) = onBindViewHolder(holder, item, position)
            override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): VH = onCreateViewHolder(parent, viewType)
        }
        bind(itemType, binder)
        return binder
    }

    fun <T> bind(itemType: Class<T>, @LayoutRes layoutId: Int, onBindViewHolder: OnBindViewHolder<T, CommonItemViewHolder>): ItemTypeViewBinder<T, CommonItemViewHolder> {
        val binder = object : AbstractItemViewBinder<T>() {
            override fun getLayoutId() = layoutId
            override fun onBindViewHolder(holder: CommonItemViewHolder, item: T, position: Int) {
                onBindViewHolder(holder, item, position)
            }
        }
        bind(itemType, binder)
        return binder
    }

    fun with(recyclerView: RecyclerView): RecyclerViewAdapter {
        recyclerView.adapter = this
        return this
    }

    fun with(recyclerView: RecyclerView, block: RecyclerViewAdapter.() -> Unit): RecyclerViewAdapter {
        this.with(recyclerView)
        this.block()
        return this
    }

    fun setItems(items: MutableList<Any>): RecyclerViewAdapter {
        this.items = items
        return this
    }

    fun appendItems(items: List<Any>): RecyclerViewAdapter {
        this.items.addAll(items)
        return this
    }

    fun appendItems(vararg items: Any): RecyclerViewAdapter {
        Collections.addAll(this.items, *items)
        return this
    }

}